clients/spark/core/src/main/scala/io/treeverse/clients/GCBackupAndRestore.scala

```diff
@@ -0,0 +1,180 @@
+      storageType: String
+  ): DataFrame = {
+    var storageNSForHadoopFS = ApiClient
+      .translateURI(URI.create(srcNamespace), storageType)
+      .normalize()
+      .toString
+    if (!storageNSForHadoopFS.endsWith("/")) {
+      storageNSForHadoopFS += "/"
+    }
+    import spark.implicits._
+    val objectsRelativePathsSeq = objectsRelativePathsDF
+      .select("address")
+      .map(_.getString(0))
+      .collect()
+      .toSeq
+
+    StorageUtils
+      .concatKeysToStorageNamespace(objectsRelativePathsSeq, storageNSForHadoopFS, storageType)
+      .toDF()
+  }
+
+  def validateAndParseHadoopConfig(
+      hc: Configuration,
+      storageType: String
+  ): Array[(String, String)] = {
+    var hadoopProps: Array[(String, String)] = null
+    storageType match {
+      case StorageUtils.StorageTypeS3 =>
+        hadoopProps = HadoopUtils.getHadoopConfigurationValues(hc, S3AccessKeyName, S3SecretKeyName)
+        if (hadoopProps == null || hadoopProps.length != 2) {
+          Console.err.println(
+            "missing required hadoop properties. " + S3AccessKeyName + " or " + S3SecretKeyName
+          )
+          System.exit(1)
+        }
+      case StorageUtils.StorageTypeAzure =>
+        hadoopProps = HadoopUtils.getHadoopConfigurationValues(hc, AzureStorageAccountKeyName)
+        if (hadoopProps == null || hadoopProps.length != 1) {
+          Console.err.println(
+            "missing required hadoop property. " + AzureStorageAccountKeyName
+          )
+          System.exit(1)
+        }
+    }
+    hadoopProps
+  }
+
+  // Construct a DistCp command of the form:
+  // `hadoop distcp -D<per storage credentials> -f absoluteAddressesTextFilePath dstNamespaceForHadoopFs`
+  // example for command format https://docs.lakefs.io/integrations/distcp.html
+  def constructDistCpCommand(
+      hadoopProps: Array[(String, String)],
+      absoluteAddressesTextFilePath: String,
+      dstNamespaceForHadoopFs: String
+  ): Array[String] = {
+    var res: Array[String] = Array()
+    for (prop <- hadoopProps) {
+      val formattedProp = "-D" + prop._1 + "=" + prop._2
+      res = res :+ formattedProp
+    }
+    // -f option copies the files listed in the file after the -f option
+    res = res :+ "-f"
+    res = res :+ absoluteAddressesTextFilePath
+    res = res :+ dstNamespaceForHadoopFs
+    res
+  }
+
+  // Find the path of the first txt file under a prefix.
+  def getTextFileLocation(prefix: String, hc: Configuration): String = {
+    val fs = FileSystem.get(URI.create(prefix), hc)
+    val dirIterator = fs.listFiles(new Path(prefix), false)
+    var textFilePath = ""
+    breakable(
+      while (dirIterator.hasNext) {
+        val curFile = dirIterator.next()
+        val curPath = curFile.getPath.toString
+        if (curPath.endsWith("txt")) {
+          textFilePath = curPath
+          break
+        }
+      }
+    )
+    textFilePath
+  }
+
+  def main(args: Array[String]): Unit = {
+    validateArgs(args)
```

Comment:
I prefer not to because then I need to validate the ars length again and I'm losing the point in the validate args function. I can get rid of it if that makes things more readable

Reply:
OK.